cachix/src/Cachix/Client/Commands.hs

```diff
@@ -238,58 +235,55 @@ pushStorePath env name storePath = retryPath $ \retrystatus -> do
              .| passthroughSizeSink fileSizeRef
              .| passthroughHashSink fileHashRef
 
-            conduitToStreaming :: S.Stream (S.Of ByteString) (ResourceT IO) ()
-            conduitToStreaming = runConduit (transPipe lift stream' .| CL.mapM_ S.yield)
         -- for now we need to use letsencrypt domain instead of cloudflare due to its upload limits
         let newClientEnv = (clientenv env) {
               baseUrl = (baseUrl (clientenv env)) { baseUrlHost = toS name <> "." <> baseUrlHost (baseUrl (clientenv env))}
             }
-        NoContent <- escalate <=< (`runClientM` newClientEnv) $ Api.createNar
-          (cachixBCClient name)
-          (contentType (Proxy :: Proxy Api.XNixNar), conduitToStreaming)
-        closeStdout
-        exitcode <- waitForStreamingProcess cph
-        -- TODO: print process stderr?
-        when (exitcode /= ExitSuccess) $ throwM $ NarStreamingError exitcode $ show cmd
-
-        narSize <- readIORef narSizeRef
-        narHashB16 <- readIORef narHashRef
-        fileHash <- readIORef fileHashRef
-        fileSize <- readIORef fileSizeRef
-
-        -- TODO: #3: implement using pure haskell
-        narHash <- ("sha256:" <>) . T.strip . toS <$> readProcess "nix-hash" ["--type", "sha256", "--to-base32", toS narHashB16] mempty
-
-        deriverRaw <- T.strip . toS <$> readProcess "nix-store" ["-q", "--deriver", toS storePath] mempty
-        let deriver = if deriverRaw == "unknown-deriver"
-                      then deriverRaw
-                      else T.drop 11 deriverRaw
-
-        references <- sort . T.lines . T.strip . toS <$> readProcess "nix-store" ["-q", "--references", toS storePath] mempty
-
-        let
-            fp = fingerprint storePath narHash narSize references
-            sig = dsign sk fp
-            nic = Api.NarInfoCreate
-              { cStoreHash = storeHash
-              , cStoreSuffix = storeSuffix
-              , cNarHash = narHash
-              , cNarSize = narSize
-              , cFileSize = fileSize
-              , cFileHash = fileHash
-              , cReferences = fmap (T.drop 11) references
-              , cDeriver = deriver
-              , cSig = toS $ B64.encode $ unSignature sig
-              }
-
-        escalate $ Api.isNarInfoCreateValid nic
-
-        -- Upload narinfo with signature
-        NoContent <- escalate <=< (`runClientM` clientenv env) $ Api.createNarinfo
-          (cachixBCClient name)
-          (Api.NarInfoC storeHash)
-          nic
-        return ()
+        void $ (`withClientM` newClientEnv)
+            (Api.createNar (cachixBCClient name) stream')
+            $ escalate >=> \NoContent -> do
```

Comment:
I don't like this combination of right-to-left and left-to-right operators.
What was wrong with the `NoContent <-` binding?

Reply:
It's a different API now, it is CPS so it's either this or the previous way you wanted to get rid of double `resp`

Reply:
:+1: